workspace/src/main/kotlin/com/cosmotech/workspace/azure/WorkspaceServiceImpl.kt

```diff
@@ -43,94 +37,101 @@ import org.springframework.stereotype.Service
       workspace.users = listOf()
       cosmosTemplate.upsert("${organizationId}_workspaces", workspace)
-
-      userIds.forEach {
-        this.eventPublisher.publishEvent(
-            UserRemovedFromWorkspace(this, organizationId, workspaceId, it))
-      }
     }
   }
 
-  override fun removeUserFromOrganizationWorkspace(
+  override fun removeUserFromWorkspace(
       organizationId: String,
       workspaceId: String,
-      userId: String
+      userMail: String
   ) {
     val workspace = findWorkspaceById(organizationId, workspaceId)
-    val workspaceUserMap =
-        workspace.users?.associateBy { it.id!! }?.toMutableMap() ?: mutableMapOf()
-    if (workspaceUserMap.containsKey(userId)) {
-      workspaceUserMap.remove(userId)
-      workspace.users = workspaceUserMap.values.toList()
-      cosmosTemplate.upsert("${organizationId}_workspaces", workspace)
-      this.eventPublisher.publishEvent(
-          UserRemovedFromWorkspace(this, organizationId, workspaceId, userId))
+    if (workspace.ownerId != getCurrentAuthenticatedUserName()) {
+      // TODO Only the owner or an admin should be able to perform this operation
+      throw CsmAccessForbiddenException("You are not allowed to update this Resource")
     }
-  }
+    workspace.users = workspace.users?.filter { it != userMail }
 
-  override fun addOrReplaceUsersInOrganizationWorkspace(
-      organizationId: String,
-      workspaceId: String,
-      workspaceUser: List<WorkspaceUser>
-  ): List<WorkspaceUser> {
-    if (workspaceUser.isEmpty()) {
-      // Nothing to do
-      return workspaceUser
-    }
+    cosmosTemplate.upsert("${organizationId}_workspaces", workspace)
+  }
 
-    val organization = organizationService.findOrganizationById(organizationId)
+  override fun addUsersInWorkspace(
+      organizationId: kotlin.String,
+      workspaceId: kotlin.String,
+      requestBody: kotlin.collections.List<kotlin.String>
+  ): List<kotlin.String> {
     val workspace = findWorkspaceById(organizationId, workspaceId)
+    if (workspace.ownerId != getCurrentAuthenticatedUserName()) {
+      // TODO Only the owner or an admin should be able to perform this operation
+      throw CsmAccessForbiddenException("You are not allowed to update this Resource")
+    }
 
-    val workspaceUserWithoutNullIds = workspaceUser.filter { it.id != null }
-    val newUsersLoaded = fetchUsers(workspaceUserWithoutNullIds.mapNotNull { it.id })
-    val workspaceUserWithRightNames =
-        workspaceUserWithoutNullIds.map { it.copy(name = newUsersLoaded[it.id]!!.name!!) }
-    val workspaceUserMap = workspaceUserWithRightNames.associateBy { it.id!! }
-
-    val currentWorkspaceUsers =
-        workspace.users?.filter { it.id != null }?.associateBy { it.id!! }?.toMutableMap()
-            ?: mutableMapOf()
-
-    newUsersLoaded.forEach { (userId, _) ->
-      // Add or replace
-      currentWorkspaceUsers[userId] = workspaceUserMap[userId]!!
+    if (requestBody.isEmpty()) {
```

Comment:
For performance reasons, I suggest returning right away if `requestBody` and `workspace.users` are identical (regardless of duplicates and regardless of the order of the elements).
This would prevent performing a useless call to `cosmosTemplate.upsert`.

Reply:
todo

Reply:
dones